Update to RxJava2

dsrees opened the pull request:
Closes #208 and updates the code to use `io.reactivex` classes instead of `rx` classes. This PR also makes #223 obsolete. Also updated the Retrofit2 RxJava adapter to the RxJava2 adapter. 

In the process I also updated Retrofit from `2.0.0-beta4` to `2.3.0` and `OkHttp` from `3.5.0` to `3.8.0`. This had some repercussions with `ChangeableBaseUrl` which I believe I resolved correctly but please check. This would also make #236 obsolete though that PR overhauls `ChangeableBaseUrl` and upgrades OkHttp to 3.9.0 while Retrofit `2.3.0` is using OkHttp `3.8.0` under the hood.

All tests are ✅

dsrees committed:
Switched from RxJava to RxJava2 and updated Retrofit 2.3.0 from -beta4

dsrees committed:
Fixed http test

vanniktech commented on app/src/main/java/com/artemzin/qualitymatters/other/DisposableSubscription.java:
```diff
@@ -1,33 +1,36 @@
+            try {
+                disposeAction.run();
+            } catch (Exception e) {
+                throw new RuntimeException(e); // RxJava2 Action throws an exception. Pass it on
```
Should we rethrow here or use the RxJavaPlugins?

dsrees replied:
Please check out the updated implementation which converts `FinishAsyncJobSubscription` into `FinishAsyncJobAction` and passes it through to `Disposables.fromAction(finishAsyncJobAction)`

vanniktech commented on app/src/main/java/com/artemzin/qualitymatters/other/DisposableSubscription.java:
```diff
@@ -1,33 +1,36 @@
-import rx.functions.Action0;
-
-public class DisposableSubscription implements Subscription {
+public class DisposableSubscription implements Disposable {
```
I feel like this could be solved with `Disposables.from` instead

vanniktech commented on dependencies.gradle:
```diff
@@ -19,12 +19,11 @@ ext.versions = [
         supportLibs                  : '23.1.1',
-        okHttp                       : '3.2.0',
-        retrofit                     : '2.0.0-beta4',
+        okHttp                       : '3.8.0',
```
there's 3.9.0

dsrees committed:
Updates after review

dsrees committed:
Remove catch of exception since it is thrown by method

dsrees commented:
Updated based on review. Not sure why the travis job is failing, but it seems to be failing for all recent PRs. Is this something I can fix in this PR?

artem-zinnatullin commented:
Sorry, I'm at KotlinConf and then will spend weekend with friends who also came to conference and then DroidconSF and work… 

I've added this to my todo list and hopefully will review in the middle of next week, apologies 😬 

I really trust @vanniktech, so you can definitely continue addressing his comments 😸

vanniktech commented:
I don't think that we need the `FinishAsyncJobSubscription` class at all. Simply using one of the Disposables should do the job:

```
Disposables.fromRunnable(Runnable run) // Construct a Disposable by wrapping a Runnable that is executed exactly once when the Disposable is disposed.
Disposables.fromAction(Action run) // Construct a Disposable by wrapping a Action that is executed exactly once when the Disposable is disposed.
```

dsrees committed:
Removed FinishAsyncJobAction in favor of lambda

dsrees commented:
* Removed `DisposableSubscription` and test
* Removed `FinishAsyncJobAction` and test
* Using single lambda expression `Disposables.fromAction(() -> asyncJobsObserver.asyncJobFinished(asyncJob))` in place of `FinishAsyncJobAction`
* Added test to ensure that the async job action was finished if the Disposable was disposed of.

vanniktech reviewed (approved):
Looks good from RxJava 2 side.

As for the `ChangeableBaseUrl`, I'll leave it up to Artem how / if that class needs a change.

Seems like CI is failing since the logs are flooded?

vanniktech commented:
@danielrees18 is this on the latest master branch? If not, rebasing should fix the build issue.

dsrees commented:
Haven't kept up with changes master, I'll do that shortly

On Dec 26, 2017 6:38 PM, "Niklas Baudy" <notifications@github.com> wrote:

> @danielrees18 <https://github.com/danielrees18> is this on the latest
> master branch? If not, rebasing should fix the build issue.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/artem-zinnatullin/qualitymatters/pull/239#issuecomment-354025211>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACwONw8MaRNCr0XhiVRBjczjebl6pKrLks5tEYNegaJpZM4QQuw_>
> .
>

vanniktech reviewed (approved):
Looks good on my side.

As for the ChangeableBaseUrl, I'll leave it up to @artem-zinnatullin how / if that class needs a change.

mibrahimdev commented:
is there any updates on this PR, interested to see Rxjava2 in this project to experiment with it.

artem-zinnatullin commented:
ProGuard config needs an update, build fails https://api.travis-ci.org/v3/job/323862434/log.txt
